tools/fix_timepoint_date_problems.php

```diff
@@ -1,531 +1,674 @@
-            $missingInstruments = array_merge($missingInstruments, $difference);
-
-            // unset vars and objects
-            unset($battery);
-            unset($neededTests);
-            unset($existingTests);
-            unset($age);
-            unset($difference);
-        }
-    } // end foreach
-
-    return $missingInstruments;
-}
-?>
+#!/usr/bin/env php
+<?php
+/**
+ * This is the tool to diagnose and correct the date problems in a candidate
+ * profile and add missing instruments to the bvl battery.
+ *
+ * Currently the tool does the following:
+ * - corrects DOB and EDC (candidate table)
+ * - corrects Date_screening, Date_visit (session table)
+ * - corrects bvl batteries by:
+ *    adding records in the flag table
+ *    adding records in the instrument tables
+ *    adding records into the feedback_bvl* tables (creates new feedback)
+ *    writes a log entry
+ *    sending automatic email notification - to be completed later
+ *
+ * The procedure to correct the problems is:
+ * Run w/ fix_date option to UPDATE the date fields in the candidate/session
+ * table (as needed)
+ * Run w/ diagnose option to get the difference in batteries
+ * Run the tool w/ add_instrument option to add missing instruments, and
+ * create new feedback
+ * Send email to PSC
+ *
+ * Ways to use this script:
+ * -- get a help screen
+ * fix_timepoint_date_problems.php help
+ *
+ * -- to fix dates (when fixing stage dates, sessionID must be specified)
+ * fix_timepoint_date_problems.php fix_date <candID> <newCorrectDate> <dob/edc>
+ * fix_timepoint_date_problems.php fix_date <candID> <newCorrectDate>
+ *                                          <screening/visit> <sessionID>
+ *
+ * -- when diagnosing a problem (if only CandID is specified, all timepoints
+ * are diagnosed
+ * fix_timepoint_date_problems.php diagnose <candID> [<sessionID>]
+ * fix_timepoint_date_problems.php diagnose <candID> [<newCorrectDate> <dob/edc/>]
+ * fix_timepoint_date_problems.php diagnose <candID> [<sessionID>
+ *                                          <newCorrectDate> <screening/visit>]
+ *
+ * -- to fix bvl battery (run only once the dates are fixed)
+ * fix_timepoint_date_problems.php add_instrument <candID> <sessionID> <test_name>
+ *
+ * PHP Version 7
+ *
+ * @category Main
+ * @package  Loris
+ * @author   Loris Team <loris-dev@bic.mni.mcgill.ca>
+ * @license  Loris license
+ * @link     https://www.github.com/aces/Loris-Trunk/
+ */
+use LORIS\StudyEntities\Candidate\CandID;
+
+set_include_path(get_include_path().":../project/libraries:../php/libraries:");
+
+// path to config file
+$configFile = "../project/config.xml";
+
+require_once __DIR__ . "/../vendor/autoload.php";
+$client = new NDB_Client();
+$client->makeCommandLine();
+$client->initialize($configFile);
+
+$db = \Database::singleton();
+
+/**
+ * HELP SCREEN
+ * display and stop processing if action=help
+ */
+if (empty($argv[1]) || $argv[1] == 'help') {
```

Comment:
Exit script after displaying help message if args are missing. Also check that argv[2], [3]..etc are set, otherwise:

```
PHP Notice:  Undefined offset: 1 in /var/www/loris/tools/fix_timepoint_date_problems.php on line 92
PHP Notice:  Undefined offset: 2 in /var/www/loris/tools/fix_timepoint_date_problems.php on line 95
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to LORIS\StudyEntities\Candidate\ValidatableIdentifier::__construct() must be of the type string, null given, called in /var/www/loris/tools/fix_timepoint_date_problems.php on line 95 and defined in /var/www/loris/src/StudyEntities/Candidate/ValidatableIdentifier.php:58
Stack trace:
#0 /var/www/loris/tools/fix_timepoint_date_problems.php(95): LORIS\StudyEntities\Candidate\ValidatableIdentifier->__construct(NULL)
#1 {main}
  thrown in /var/www/loris/src/StudyEntities/Candidate/ValidatableIdentifier.php on line 58
```

Reply:
Thank you @jesscall for all you suggestions. I will have a look shortly.